src/provider/provider_level_zero.c

```diff
@@ -0,0 +1,240 @@
+
+#include <umf.h>
+#include <umf/memory_provider_ops.h>
+#include <umf/providers/provider_level_zero.h>
+
+#include "base_alloc_global.h"
+#include "utils_common.h"
+#include "utils_concurrency.h"
+#include "utils_sanitizers.h"
+
+typedef struct ze_memory_provider_t {
+    ze_context_handle_t context;
+    ze_device_handle_t device;
+    ze_memory_type_t memory_type;
+    ze_host_mem_alloc_desc_t host_mem_alloc_desc;
+    ze_device_mem_alloc_desc_t device_mem_alloc_desc;
+} ze_memory_provider_t;
+
+typedef struct ze_ops_t {
+    ze_result_t (*zeMemAllocDevice)(ze_context_handle_t,
+                                    const ze_device_mem_alloc_desc_t *, size_t,
+                                    size_t, ze_device_handle_t, void *);
+    ze_result_t (*zeMemFree)(ze_context_handle_t, void *);
+} ze_ops_t;
+
+static ze_ops_t g_ze_ops;
+static UTIL_ONCE_FLAG ze_is_initialized = UTIL_ONCE_FLAG_INIT;
+static bool Init_ze_global_state_failed;
+
+static void init_ze_global_state(void) {
+    // check if Level Zero shared library is already loaded
+    *(void **)&g_ze_ops.zeMemAllocDevice =
+        dlsym(RTLD_DEFAULT, "zeMemAllocDevice");
+    *(void **)&g_ze_ops.zeMemFree = dlsym(RTLD_DEFAULT, "zeMemFree");
+
+    if (!g_ze_ops.zeMemAllocDevice || !g_ze_ops.zeMemFree) {
+        fprintf(stderr, "Required Level Zero symbols not found.\n");
+        Init_ze_global_state_failed = true;
+    }
+}
+
+enum umf_result_t ze_memory_provider_initialize(void *params, void **provider) {
+    if (provider == NULL || params == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    level_zero_memory_provider_params_t *ze_params =
+        (level_zero_memory_provider_params_t *)params;
+
+    util_init_once(&ze_is_initialized, init_ze_global_state);
+    if (Init_ze_global_state_failed) {
+        fprintf(stderr, "Loading Level Zero symbols failed\n");
+        return UMF_RESULT_ERROR_UNKNOWN;
+    }
+
+    if (ze_params->memory_type != UMF_MEMORY_TYPE_DEVICE) {
+        // TODO support other memory types
+        return UMF_RESULT_ERROR_NOT_SUPPORTED;
+    }
+
+    ze_memory_provider_t *ze_provider =
+        umf_ba_global_alloc(sizeof(ze_memory_provider_t));
+    if (!ze_provider) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    ze_provider->context = ze_params->level_zero_context_handle;
+    ze_provider->device = ze_params->level_zero_device_handle;
+    ze_provider->memory_type = ze_params->memory_type;
+    ze_provider->host_mem_alloc_desc = (ze_host_mem_alloc_desc_t){
+        ZE_STRUCTURE_TYPE_HOST_MEM_ALLOC_DESC,
+        NULL, // pointer to an extension-specific structure
+        ze_params->level_zero_host_mem_alloc_flags};
+    ze_provider->device_mem_alloc_desc = (ze_device_mem_alloc_desc_t){
+        ZE_STRUCTURE_TYPE_DEVICE_MEM_ALLOC_DESC,
+        NULL, // pointer to an extension-specific structure
+        ze_params->level_zero_device_mem_alloc_flags,
+        ze_params->level_zero_device_local_mem_ordinal};
```

Comment:
I do not understand where it is used and why it is needed?

Reply:
you have to pass ze_device_mem_alloc_desc_t struct as an argument to zeMemAllocDevice

Reply:
right, but we can create it on stack inside `ze_memory_provider_alloc` function. Furthermore, `device_mem_alloc_desc` is needed only when `ze_provider->memory_type` equals to the device memory - it is another argument not to make it part of the `ze_params` object.